File: js/handlers/mongoFilterHandler.js
const { MongoFilterBuilder } = require('filtex-js');
const { ProjectFilter } = require('../filters/projectFilter');
const { MongoStorage } = require('../storages/mongoStorage');
const {MongoExpression} = require("filtex-js/dist/builders/mongo/types/mongoExpression");

class MongoFilterHandler {
    async handleAsync(req, res) {
        try {
            let expression;

            switch (req.query.type) {
                case 'text':
                    if (!req.body.query || req.body.query === '') {
                        break;
                    }

                    ProjectFilter.validateFromText(req.body.query);
                    expression = ProjectFilter.expressionFromText(req.body.query);
                    break;
                case 'json':
                    if (!req.body.query) {
                        break;
                    }

                    ProjectFilter.validateFromJson(JSON.stringify(req.body.query));
                    expression = ProjectFilter.expressionFromJson(JSON.stringify(req.body.query));
                    break;
                default:
                    throw new Error('invalid type');
            }

            let mongoFilter;

            if (expression) {
                mongoFilter = new MongoFilterBuilder().build(expression);
            } else {
                mongoFilter = new MongoExpression({});
            }

            const results = await new MongoStorage().Query(mongoFilter.condition);

            res.status(200).json(results);
        } catch (ex) {
            res.status(500).send(ex.message);
        }
    }
}

module.exports = { MongoFilterHandler };

File: go/handlers/mongo_filter_handler.go
package handlers

import (
	"encoding/json"
	"filtex-go-example/filters"
	"filtex-go-example/storages"
	"github.com/filtex/filtex-go/builders/mongo"
	"github.com/filtex/filtex-go/builders/mongo/types"
	"github.com/filtex/filtex-go/expressions"
	"go.mongodb.org/mongo-driver/bson"
	"net/http"
)

func MongoFilterHandler(w http.ResponseWriter, r *http.Request) {
	projectFilter, err := filters.ProjectFilter()
	if err != nil {
		http.Error(w, "Error generating Filtex instance", http.StatusInternalServerError)
		return
	}

	var requestBody map[string]interface{}
	decoder := json.NewDecoder(r.Body)
	err = decoder.Decode(&requestBody)
	if err != nil {
		http.Error(w, "Error decoding JSON request", http.StatusBadRequest)
		return
	}

	var expression expressions.Expression

	switch r.URL.Query().Get("type") {
	case "text":
		if requestBody["query"] == nil {
			break
		}

		query, ok := requestBody["query"].(string)
		if !ok {
			http.Error(w, "Invalid request format", http.StatusBadRequest)
			return
		}

		if query == "" {
			break
		}

		err = projectFilter.ValidateFromText(query)
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}

		expression, err = projectFilter.ExpressionFromText(query)
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}
	case "json":
		if requestBody["query"] == nil {
			break
		}

		query, err := json.Marshal(requestBody["query"])
		if err != nil {
			http.Error(w, "Invalid request format", http.StatusBadRequest)
			return
		}

		err = projectFilter.ValidateFromJson(string(query))
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}

		expression, err = projectFilter.ExpressionFromJson(string(query))
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}
	default:
		http.Error(w, "invalid type", http.StatusBadRequest)
		return
	}

	var mongoFilter *types.MongoExpression

	if expression != nil {
		mongoFilter, err = mongo.NewMongoFilterBuilder().Build(expression)
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}
	} else {
		mongoFilter = &types.MongoExpression{
			Condition: bson.M{},
		}
	}

	results, err := storages.MongoStorage{}.Query(mongoFilter.Condition)
	if err != nil {
		http.Error(w, "could not be fetched", http.StatusBadRequest)
		return
	}

	jsonResponse, err := json.Marshal(results)
	if err != nil {
		http.Error(w, "Error encoding JSON response", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")

	w.Write(jsonResponse)
}
